docs/alternativesemantics.md

```diff
@@ -0,0 +1,202 @@
+
+### Type Rules
+
+Aliasing assignment generates a fresh region
+if it follows an `iso`, if it follows an `imm` then it is in the immutable region, otherwise it is in the same region. 
+
+Read versus Write types for a field
+```
+   field_r(tau1 | tau2, f) = field_r(tau1, f) | field_r(tau2, f)
+   field_r(tau1 & tau2, f) = field_r(tau1, f) & field_r(tau2, f)
+   field_w(tau1 | tau2, f) = field_r(tau1, f) & field_r(tau2, f)
+   field_w(tau1 & tau2, f) = field_r(tau1, f) & field_r(tau2, f)
+```
+]
+
+```
+field_R(f, tau_x) = tau_f
+fresh r
+-------------------------------------
+G |- z = x.f -| G, z: tau_x ->r tau_f
+```
+
+Extraction uses a fresh region
+```
+G(x) : tau_x
+G(y) : tau_y
+field_w(f, tau_x) = tau_f_w
+field_r(f, tau_x) = tau_f_r
+exists r'. tau_y <: tau_x <~r' tau_f_w
+fresh r
+--------------------------------------
+G |- z = (x.f = y) -| G, z: tau_x <~r tau_f_r
+```
+Is y still accessible?  If it was an iso, then no.  But otherwise, yes.  So remove y, if we can prove it was not an iso?
+
+
+Method call?
+
+No longer has to invalidate any regions, may consume locals of type `iso`.
+```
+  foo (x: iso&C) {...}
+
+  var y = new C
+  foo (y)
+  foo (y) /// Error `y` was invalidate by previous line.
+```
+
+
+
+##  Work required
+
+* Runtime support for delaying reclaimation
+* Runtime support for delaying message send
+* Feasability work for optimisations 
+    - What can we optimise?
+    - What user annotation might we provide?
+    - 
+* Type system implementation 
+* Optimisation
+
+## Concerns
+
+### Additional root sets
+
+Not tracking regions and subregions precisely means we may need to be more careful with codegen for the additional root set.  For instance, we can use have the same region multiple times, which means we use the region additional roots in a globally stack based way.  I.e.
+```
+  r1.add_root(a)
+  r2.add_root(b)
+  ...
+  r1.remove_root(a)
+  ...
+  r2.remove_root(b)
+```
+Is not allowed as if r1 and r2 are the same region, then we will not be following the stack discipline.
+
+**[This may already be the case with cown aliasing, where we don't know if two cowns are the same or different.]**
+
+### Performance
+
+This approach can cause more memory and less concurrency of messages.  This may be bad for performance.
+
+Using compiler optimisations and global analysis could mitigate the default performance by finding cases where messages can be sent promptly.
+
+This however leads to hard to predict performance for the programmer.  We can mitigate this by providing annotations the programmer provides to guarantee certain properties must be true for optimisation, but then we are effectively implementing the complex type system.  This may still be acceptable as the barrier to entry is lower, and the performance work will require a more specialist view.  We can also potentially forbid more in the perf setting, allowing programmers to trade off flexibity of implementation with performance.
```

Comment:
My worry with type annotations is that people can find multiple special cases that provide more performance and hard-code those in their programs (and evangelise other programmers to do the same), which will make it harder for us to improve the semantics later on, with hindsight.

At such an early stage, I'd pay the performance cost and keep all annotations "private" (ie. explicitly not part of the language, but available for the compiler) and try to clean them up before we start asking people to use Verona.

Reply:
Agreed, this is forward looking, but we want to have in mind we will be doing this.  Part of the reason for making the semantics allow for the messages send at any point after the references into the region have gone away. We need to leave space in the semantics for optimisation.